rust/connlib/tunnel/src/tests.rs

```diff
@@ -74,103 +108,679 @@ impl StateMachineTest for TunnelTest {
+                );
             }
-            Transition::SendICMPPacketToIp4Resource { src, r_idx } => {
+            Transition::SendICMPPacketToIp4Resource { r_idx } => {
                 let dst = ref_state.sample_ipv4_cidr_resource_dst(&r_idx);
 
-                state.send_icmp_packet_client_to_gateway(src, dst);
+                buffered_transmits
+                    .extend(state.send_icmp_packet_client_to_gateway(state.client.tunnel_ip4, dst));
             }
-            Transition::SendICMPPacketToIp6Resource { src, r_idx } => {
+            Transition::SendICMPPacketToIp6Resource { r_idx } => {
                 let dst = ref_state.sample_ipv6_cidr_resource_dst(&r_idx);
 
-                state.send_icmp_packet_client_to_gateway(src, dst);
+                buffered_transmits
+                    .extend(state.send_icmp_packet_client_to_gateway(state.client.tunnel_ip6, dst))
             }
             Transition::Tick { millis } => {
                 state.now += Duration::from_millis(millis);
-                state.utc_now += Duration::from_millis(millis);
-                state.client.handle_timeout(state.now);
-                state.gateway.handle_timeout(state.now, state.utc_now);
             }
         };
+        state.advance(ref_state, &mut buffered_transmits);
 
-        for (time, expected_event) in &ref_state.client_new_events {
-            assert_eq!(time, &state.now);
-            assert_eq!(expected_event, &state.client.poll_event().unwrap());
-        }
+        // Assert: Check that our actual state is equivalent to our expectation (the reference state).
+        assert_eq!(
+            state.gateway_received_icmp_packets,
+            ref_state.gateway_received_icmp_packets
+        );
+        assert!(buffered_transmits.is_empty()); // Sanity check to ensure we handled all packets.
 
-        // TODO: Drain `poll_transmit` and execute it.
+        state
+    }
+}
+
+/// Implementation of our reference state machine.
+///
+/// The logic in here represents what we expect the [`ClientState`] & [`GatewayState`] to do.
+/// Care has to be taken that we don't implement things in a buggy way here.
+/// After all, if your test has bugs, it won't catch any in the actual implementation.
+impl ReferenceStateMachine for ReferenceState {
+    type State = Self;
+    type Transition = Transition;
+
+    fn init_state() -> proptest::prelude::BoxedStrategy<Self::State> {
+        (
+            sim_node_prototype(client_id()),
+            sim_node_prototype(gateway_id()),
+            sim_relay_prototype(),
+            Just(Instant::now()),
+            Just(Utc::now()),
+        )
+            .prop_filter(
+                "client and gateway priv key must be different",
+                |(c, g, _, _, _)| c.state != g.state,
+            )
+            .prop_filter(
+                "viable network path must exist",
+                |(client, gateway, relay, __, _)| {
+                    if client.ip4_socket.is_some()
```

Comment:
can you add a `TODO` here like in preconditions?

Reply:
I think this may actually be redundant, now that I've made the relays always dual-stack in the tests.